### ISSUE TITLE
Cannot reproduce the results

### ISSUE THREAD

**Issue by PaladinEE15** [caslab-vt/SARNet#3]
Hello, I ran the following commands as mentioned in Readme.md :
python train.py --policy-grad reinforce --env-type ic3net --scenario traffic_junction --num-adversaries 6 --key-units 32 --value-units 32 --query-units 32 --len-traj-update 10 --encoder-model LSTM --max-episode-len 20 --add_rate_min 0.3 --num-env 200 --num-episodes 1000000
and after training for a long time, the output results were still bad:
 steps: 19752000, episodes: 987600.0, mean episode success: 0.3341001923855812, time: 28891.135
Is there something wrong with my settings?

**Comment by murtazarang** [comment]
Can you please try the following:

`python train.py --env-type ic3net --scenario traffic_junction --policy-grad reinforce --num-adversaries 6 --adv-test SARNET --gpu-device 0 --exp-name SAR-TJ6 --max-episode-len 20 --num-env 16 --dim 6 --add_rate_min 0.1 --add_rate_max 0.3 --curr_start 250 --curr_end 1250 --num-episodes 3000
`

I'll update the README to show the current code in a couple of days.

**Comment by PaladinEE15** [comment]
> Can you please try the following:
> 
> `python train.py --env-type ic3net --scenario traffic_junction --policy-grad reinforce --num-adversaries 6 --adv-test SARNET --gpu-device 0 --exp-name SAR-TJ6 --max-episode-len 20 --num-env 16 --dim 6 --add_rate_min 0.1 --add_rate_max 0.3 --curr_start 250 --curr_end 1250 --num-episodes 3000 `
> 
> I'll update the README to show the current code in a couple of days.

The results are still bad. I noticed that the training is completed after a few minutes, and that's obviously not enough for a MARL task. Maybe the batch-size is set too small (default: 256). I tried the released codes of IC3Net paper, where each subprocess (16 in total) runs 500 episodes before a joint gradient update. However, increasing the batch-size still cannot lead to better performance here.

**Comment by murtazarang** [comment]
Can you please quantify your results, and for which communication architectures so that I can troubleshoot them?

**Comment by PaladinEE15** [comment]
I run the following command:
python train.py --env-type ic3net --scenario traffic_junction --policy-grad reinforce --num-adversaries 6 --adv-test SARNET --gpu-device 0 --exp-name SAR-TJ6 --max-episode-len 20 --num-env 16 --dim 6 --add_rate_min 0.1 --add_rate_max 0.3 --curr_start 250 --curr_end 1250 --num-episodes 3000
and get good results (success~0.8). However, the add_rate (rate at which to add car) of the Traffic Junction environment remains  0.1. This is because the environment updates its add_rate corresbonding to epoch, and 3000 episodes= 3000/16 epochs = 188 epochs < 250. This means the environment doesn't increase its add_rate during test time.

Then I change the command to make sure that the environment increase its add_rate currectly:
python train.py --env-type ic3net --scenario traffic_junction --policy-grad reinforce --num-adversaries 6 --adv-test SARNET --gpu-device 0 --exp-name SAR-TJ6 --max-episode-len 20 --num-env 16 --dim 6 --add_rate_min 0.1 --add_rate_max 0.3 --curr_start 16 --curr_end 78 --num-episodes 3000
(16 = int(250/16), 78 = int(1250/16) )
and get bad results: 
steps: 59840, episodes: 2992.0, mean episode success: 0.3114583333333333, time: 726.898

**Comment by murtazarang** [comment]
There was a bug in the code upon release. Thank you for pointing it out. I will push out an update within the next 24 hours to fix that.

**Comment by murtazarang** [comment]
I fixed the published code for correct training for the REINFORCE algorithm. Please reopen the issue if needed. And thank you again for pointing out the error.

**Comment by murtazarang** [comment]
[https://github.com/caslab-vt/SARNet/issues/4](https://github.com/caslab-vt/SARNet/issues/4[](url))

Continuing the thread here. It might be because of the Tensorflow upgrades? Have you cross checked dropout rates for SARNet while upgrading?

I am currently training the specific ones for N=6 agents, and will post results soon once completed with network models, and screenshots.

**Comment by murtazarang** [comment]
I ran a few experiments and here are the results for Traffic Junction:

**SARNet N=6 Easy No Curriculum**

`python train.py --env-type ic3net --scenario traffic_junction --policy-grad reinforce --num-adversaries 6 --adv-test SARNET --gpu-device 0 --exp-name SAR-TJ6-NoCurrLr --max-episode-len 20 --num-env 50 --dim 6 --add_rate_min 0.3 --add_rate_max 0.3 --curr_start 250 --curr_end 1250 --num-episodes 500000 --batch-size 500 --difficulty easy --vision 0 --batch-size 500`

![SARNet-TJ6-No Curriculum](https://user-images.githubusercontent.com/18159151/133898226-11e7e2f1-d667-43e9-9a57-9bc8849bb8bc.png)

**SARNet N=6 Easy with Curriculum**

`python train.py --env-type ic3net --scenario traffic_junction --policy-grad reinforce --num-adversaries 6 --adv-test SARNET --gpu-device 1 --exp-name SAR-TJ6-CurrLr --max-episode-len 20 --num-env 50 --dim 6 --add_rate_min 0.1 --add_rate_max 0.3 --curr_start 250 --curr_end 1250 --num-episodes 500000 --batch-size 500 --difficulty easy --vision 0 --batch-size 500`

![SARNet-TJ6-Curriculum](https://user-images.githubusercontent.com/18159151/133898233-3aa1551d-9ebd-43cc-92dc-13e182582ea5.png)

**IC3Net N=6 No Curriculum from the [IC3Net ](https://github.com/IC3Net/IC3Net )repository** 
Note: I couldn't get their multiprocessing to work so I set nprocesses 0 as a Single Trainer.

`python main.py --env_name traffic_junction --nagents 6 --nprocesses 0 --num_epochs 2000 --hid_size 128 --detach_gap 10 --lrate 0.001 --dim 6 --max_steps 20 --ic3net --vision 0 --recurrent --add_rate_min 0.3 --add_rate_max 0.3 --curr_start 250 --curr_end 1250 --difficulty easy`

![IC3Net TJ6 No Curriculum](https://user-images.githubusercontent.com/18159151/133898254-ef7d58b1-05e5-4bc1-83d2-b0765f2c2575.png)

The results vary for TJ6 as compared to the published paper due to the error, and this will be updated on an Arxiv submission in the coming weeks. However, as we still observe that SARNet learns from a No Curriculum

Also including a N=10 No Curriculum training here, but unfortunately I did not take screen shots of it before closing the window. These are from the training values file.

**SARNet N=10 Medium No Curriculum**
`python train.py --env-type ic3net --scenario traffic_junction --policy-grad reinforce --num-adversaries 10 --adv-test SARNET --gpu-device 0 --exp-name SAR-TJ10-NoCurrLr --max-episode-len 40 --num-env 15 --dim 14 --add_rate_min 0.05 --add_rate_max 0.05 --curr_start 250 --curr_end 1250 --num-episodes 500000 --batch-size 500 --difficulty medium --vision 0 --batch-size 500`

> steps: 13222800, episodes: 330570.0, mean episode success: 0.903888888888889, time: 37888.13
> 
> steps: 13223400, episodes: 330585.0, mean episode success: 0.9063888888888888, time: 37894.516
> 
> steps: 13224000, episodes: 330600.0, mean episode success: 0.9163888888888889, time: 37900.878
> 
> steps: 13224600, episodes: 330615.0, mean episode success: 0.9052777777777778, time: 37907.27
> 
> steps: 13225200, episodes: 330630.0, mean episode success: 0.9019444444444445, time: 37913.664
> 
> steps: 13225800, episodes: 330645.0, mean episode success: 0.9180555555555554, time: 37920.096
> 
> steps: 13226400, episodes: 330660.0, mean episode success: 0.8852777777777779, time: 37926.41
> 
> steps: 13227000, episodes: 330675.0, mean episode success: 0.877222222222222, time: 37932.719
> 
> steps: 13227600, episodes: 330690.0, mean episode success: 0.8644444444444443, time: 37939.114
> 
> steps: 13228200, episodes: 330705.0, mean episode success: 0.8533333333333333, time: 37945.465
> 
> steps: 13228800, episodes: 330720.0, mean episode success: 0.8308333333333333, time: 37951.837
> 
> steps: 13229400, episodes: 330735.0, mean episode success: 0.8077777777777777, time: 37958.178
> 
> steps: 13230000, episodes: 330750.0, mean episode success: 0.8191666666666666, time: 37964.548
> 
> steps: 13230600, episodes: 330765.0, mean episode success: 0.8424999999999999, time: 37971.009
> 
> steps: 13231200, episodes: 330780.0, mean episode success: 0.8538888888888888, time: 37976.042
> 
> steps: 13231800, episodes: 330795.0, mean episode success: 0.8697222222222223, time: 37982.458
> 
> steps: 13232400, episodes: 330810.0, mean episode success: 0.8947222222222221, time: 37988.823
> 
> steps: 13233000, episodes: 330825.0, mean episode success: 0.9025000000000001, time: 37995.191
> 
> steps: 13233600, episodes: 330840.0, mean episode success: 0.9155555555555557, time: 38001.576
> 
> steps: 13234200, episodes: 330855.0, mean episode success: 0.9044444444444446, time: 38007.958
> 
> steps: 13234800, episodes: 330870.0, mean episode success: 0.9161111111111113, time: 38014.42

**Comment by PaladinEE15** [comment]
Thank you for your detailed experiments! That help a lot. Maybe I should try training with more episodes
